lib/file_mutex.rb

```diff
@@ -41,37 +44,43 @@ def initialize(path)
   # @return The result of the block.
 
   def lock!(timeout_duration = DEFAULT_TIMEOUT)
-    result = nil
+    # Beware: this is a block call until the specified block is executed.
+    loop do
+      # Tries to open a file. Creates the file if the file does not exist.
+      File.open(@path, File::CREAT|File::WRONLY, 0644) do |f|
+        # Tries to acquire exclusive access to the file without waiting.
+        if f.flock(File::LOCK_EX | File::LOCK_NB)
+          # We have acquired exclusive access to the file.
+          # Executes the block and releases the exclusive access (unlocking the file).
+          begin
+            f.puts contents
+            f.flush
 
-    File.open(@path, File::CREAT|File::EXCL|File::WRONLY, 0644) do |f|
-      f.puts contents
-      f.flush
-      Timeout.timeout(timeout_duration) do
-        begin
-          result = yield
-        ensure
-          unlock!
+            Timeout.timeout(timeout_duration) do
+              return yield
+            end
+          ensure
+            unlock!
+          end
         end
       end
+
+      # Failed to acquire exclusive access to the file. It means someone
+      # else is using the file. Waits a little and tries again.
+      sleep RETRY_INTERVAL
     end
-    return result
-  rescue Errno::EEXIST
-    sleep 1
-    retry
-  ensure
-    unlock!
   end
   alias synchronize lock!
 
+  private
```

Comment:
why did you make this private?

Reply:
I made the `unlock` private because it must never called outside of `lock`.

The implementation calls `unlock` when leaving `lock`. It can't be used as `lock; do-sth; unlock`. The only to use it is `synchronize { do-sth }`.